### PR TITLE
clustermesh-apiserver: rework identities, endpoints and nodes synchronization to improve performance

### PR DESCRIPTION
**Pull request by giorio94**
This PR modifies the clustermesh-apiserver implementation to use the Resource[T] abstraction to retrieve CiliumIdentities, CiliumNodes and CiliumEndpoints from Kubernetes, rather than a vanilla informer, and synchronize them through the WorkqueueSyncStore. This ensures consistent behavior across all resource types (i.e., the retry logic is implemented by the SyncStore, rather than through the Resource[T] underlying workqueue).

This approach has several advantages, including:
* Enables retrying kvstore operations in case of temporary failures;
* Performs coalescencing in case of multiple updates for the same key;
* Simplifies the overall logic, since conversions and sanity checks are already performed by the Resource[T] abstraction.

Please, refer to the commit descriptions for more details about the individual changes.

Related: #25562

```release-note
clustermesh-apiserver: rework identities, endpoints and nodes synchronization to improve performance
```

### CODE REVIEW

**Comment by giorio94** [comment]
Force pushed to rebase onto main and resolve conflicts.

**Comment by giorio94** [comment]
k8s-1.26-kernel-net-next hit known flake https://github.com/cilium/cilium/issues/25411. I'll wait for the reviews before rerunning, since it is likely tests will need to be rerun anyways.

File: clustermesh-apiserver/main.go
Comment (marseel) [inline clustermesh-apiserver/main.go:446]: Shouldn't we also delete ip from stale in case of an error here?

Comment (giorio94) [reply clustermesh-apiserver/main.go:446]: I'm not sure about that. With the current implementation, a possibly existing entry for that IP would be removed as considered stale. I personally feel that this is correct as behavior, given that we should have replaced it with a different one.

Comment (marseel) [reply clustermesh-apiserver/main.go:446]: ok, I guess it doesn't matter too much - now we are getting error here only when encoding fails.

Comment (giorio94) [reply clustermesh-apiserver/main.go:446]: Yeah, exactly. An error *should* never be returned there.

File: clustermesh-apiserver/vmmanager.go
Comment (tklauser) [inline clustermesh-apiserver/vmmanager.go]: Nit: looks like after this change all callers of `(*CachingIdentityAllocator). InitIdentityAllocator` set the `identityStore` argument to `nil`. So might as well remove it?

Comment (giorio94) [reply clustermesh-apiserver/vmmanager.go]: I've opened a separate PR to complete the cleanup, to avoid increasing the scope of this one which is already fairly large: https://github.com/cilium/cilium/pull/25594

File: clustermesh-apiserver/main.go
Comment (marseel) [inline clustermesh-apiserver/main.go]: nit: It's the same as one below, we can remove it.

**Review by marseel — APPROVED** [review]
Looks good, small nit only. I haven't reviewed the `WithTransform` part as it seems it comes from another PR.

File: clustermesh-apiserver/main.go
Comment (giorio94) [reply clustermesh-apiserver/main.go]: Fixed, thanks!

**Comment by giorio94** [comment]
> Looks good, small nit only. I haven't reviewed the `WithTransform` part as it seems it comes from another PR.

Yes, the `WithTransform` comes from https://github.com/cilium/cilium/pull/23977. I'll perform a rebase once that gets merged.

**Comment by giorio94** [comment]
Force-pushed to add one commit making the `clusterName` parameter mandatory in `NewWorkqueueSyncStore`, for symmetry with this Yutaro's suggestion: https://github.com/cilium/cilium/pull/25499#discussion_r1206178794

**Comment by giorio94** [comment]
Force-pushed to rebase onto main and fix conflicts.

**Review by jibi — APPROVED** [review]

**Review by nebril — APPROVED** [review]

**Review by tklauser — APPROVED** [review]

**Comment by giorio94** [comment]
Rebased onto main to fix conflicts. I've also added the last commit to drop the temporary backend global variable introduced in https://github.com/cilium/cilium/pull/25554, since no longer necessary.

**Comment by giorio94** [comment]
Rebased onto main to fix conflicts.

**Pull request merged**
